tests/integration/helpers.py

```diff
@@ -0,0 +1,179 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Helpers for Jenkins-k8s-operator charm integration tests."""
+import textwrap
+import typing
+
+import jenkinsapi.jenkins
+import requests
+from juju.unit import Unit
+from pytest_operator.plugin import OpsTest
+
+import jenkins
+
+
+async def install_plugins(
+    ops_test: OpsTest,
+    unit: Unit,
+    jenkins_client: jenkinsapi.jenkins.Jenkins,
+    plugins: typing.Iterable[str],
+) -> None:
+    """Install plugins to Jenkins unit.
+
+    Args:
+        ops_test: The Ops testing fixture.
+        unit: The Jenkins unit to install plugins to.
+        jenkins_client: The Jenkins client of given unit.
+        plugins: Desired plugins to install.
+    """
+    plugins = tuple(plugin for plugin in plugins if not jenkins_client.has_plugin(plugin))
+    if not plugins:
+        return
+
+    returncode, stdout, stderr = await ops_test.juju(
+        "ssh",
+        "--container",
+        "jenkins",
+        unit.name,
+        "java",
+        "-jar",
+        f"{jenkins.EXECUTABLES_PATH / 'jenkins-plugin-manager-2.12.11.jar'}",
+        "-w",
+        f"{jenkins.EXECUTABLES_PATH / 'jenkins.war'}",
+        "-d",
+        str(jenkins.PLUGINS_PATH),
+        "-p",
+        " ".join(plugins),
+    )
+    assert (
+        not returncode
+    ), f"Non-zero return code {returncode} received, stdout: {stdout} stderr: {stderr}"
+    # When there are connectivity issues, it prints in stderr and if not it prints in stdout.
+    assert any(
+        ("Done" in stdout, "Done" in stderr)
```

Comment:
If `Done` is in stderr, did the plugins actually install? Or do the connectivity issues cause the install to fail?

Reply:
Sometimes the "Done" is printed on the stderr if there have been connectivity issues and it tries some other mirrors/links. As long as "Done" is in the string, it is installed correctly.

Reply:
perhaps clarify that in the comment? The current comment doesn't seem to include the details that it tries other mirrors/ links

Reply:
Added comment, thank you!